java/vector/src/main/codegen/templates/DenseUnionVector.java

```diff
@@ -714,8 +714,22 @@ public int getBufferSizeFor(final int count) {
     if (count == 0) {
       return 0;
     }
+
+    int[] legCounts = new int[typeFields.length];
+    for (int i = 0; i < count; i++) {
+      byte tid = getTypeId(i);
+      legCounts[tid]++;
+    }
+
+    long legBytes = 0;
+    byte tid = 0;
+    for (ValueVector v : internalStruct) {
+      legBytes += (long)v.getBufferSizeFor(legCounts[tid]);
+      tid++;
+    }
+
     return (int) (count * TYPE_WIDTH + (long) count * OFFSET_WIDTH
-        + DataSizeRoundingUtil.divideBy8Ceil(count) + internalStruct.getBufferSizeFor(count));
+        + DataSizeRoundingUtil.divideBy8Ceil(count) + legBytes);
```

Comment:
I think `DataSizeRoundingUtil.divideBy8Ceil(count)` is a mistake: union vectors don't have a validity bitmap, so they don't need any buffer for it.